### ISSUE TITLE
"Word boundary" in regular expression does not work as expected

### ISSUE THREAD

**Issue by martingregoire** [dan-lovelace/word-replacer-max#5]
I noticed that the "word boundary" metacharacter `\b` in a regular expression does not work as expected.

Given this rule:

```json
"matchers": [
  {
    "active": true,
    "identifier": "...",
    "queries": ["(\\d+),5\\b"],
    "queryPatterns": ["regex"],
    "replacement": "$1,test"
  }
]
```

And this page content:

```html
<div>Hours: 7,5</div>
<div>Hours: 8,55</div>
```

I would expect the rule to match the first div, and replace it with `Hours: 7,test`. The second div should not be matched.
However the first div is not matched, which seems to be because of the `\b`.

When removing this from the rule:

```json
"matchers": [
  {
    "active": true,
    "identifier": "...",
    "queries": ["(\\d+),5"],
    "queryPatterns": ["regex"],
    "replacement": "$1,test"
  }
]
```

Given the same page content, this will replace *both* divs, which in my use case is unwanted behavior. I would need to only replace the exactly matching first div.

I tried using the "Match Whole Word" button (which adds `"wholeWord"` to `"queryPatterns"`), but this also does not work.

Is there a way to use a "word boundary" in a regular expression?

**Comment by martingregoire** [comment]
As a workaround this seems to work: use a "negative lookbehind" instead of the "word boundary".

So `(\d+),5(?!\d)` (read as "match `,5` followed by any non-digit") instead of `(\d+),5\b`.

Or as a rule in the exported JSON:

```json
"matchers": [
  {
    "active": true,
    "identifier": "...",
    "queries": ["(\\d+),5(?!\\d)"],
    "queryPatterns": ["regex"],
    "replacement": "$1,test"
  }
]
```

**Comment by dan-lovelace** [comment]
Hey @martingregoire thanks for taking the time to look into a workaround that works for you.

**Comment by martingregoire** [comment]
Hi @dan-lovelace, maybe this is something in the end that could be improved?

The page where I need to replace text renders a calendar, which contains some numbers that represent "hours worked on this day". Please see this screenshot as reference:

![calendar](https://github.com/user-attachments/assets/f09129a2-1372-47f1-a0d5-1ce907b4054c)

When creating a regular expression rule for this example, like this: `(\d+),75`
It would find this text, and replace it.

However I need to add the "word boundary" to make sure that only this specific text is found, and nothing else (imagine that there could be a `6,751234`as well somewhere on this page).

So the initial idea of using `(\d+),75\b` did not work out, but `(\d+),75(?!\d)` did, as written above.

But as it turns out, on this specific page (see screenshot) this "negative lookbehind" does not work as well. The occurrence is not found.

I think the reason is the use of `.textContent`. When using `body.textContent`, the page shown in the screenshot is returned as this (excerpt):

```
MondayTuesdayWednesdayThursdayFridaySaturdaySunday1Hours: 82Hours: 83Hours: 84Hours: 85Hours: 6,7567
```

(I replaced the actual values, except for the last one, with `8`)

What we see here is that the day numbers (in the screenshot: 5 and 6) are "merged", without any whitespace, to the "Hours: ..." string. So this part: `5Hours: 6,7567`does actually mean: "on day 5, 6,75 hours were worked; on day 6, nothing; on day 7, nothing". The day numbers are just concatenated, as there is no content at all in those calendar cells.

This makes it hard to find regular expressions that match only these specific "hours worked" strings, but nothing else.

So my idea is to use `.innerText` instead of `.textContent`. Could that work?

Please see https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText:

> **Note**: `innerText` is easily confused with [Node.textContent](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent), but there are important differences between the two. Basically, `innerText` is aware of the rendered appearance of text, while `textContent` is not.

Using `body.innerText` would yield a string like this:

```
Monday
Tuesday
Wednesday
Thursday
Friday
Saturday
Sunday
1
Hours: 8
2
Hours: 8
3
Hours: 8
4
Hours: 8
5
Hours: 6,75
6
7
```

So it would add some linebreaks in between the single text nodes, and this would make it easier to create regular expressions.

WDYT, could this work? And sorry for the wall of text.

**Comment by dan-lovelace** [comment]
Hey @martingregoire - I've started looking into this. I remember going with `textContent` early on because it's more performant than `innerText` but I can understand it being problematic in this case.

One key difference between the two is `innerText` evaluates the visibility of text. The tests I've written thus far did not take this into account so I'm working on migrating them to use a real browser environment where the test elements will be rendered on-screen. This will be a big improvement to the tests in general so I appreciate you bringing it up. Hope to have an answer to this question soon.

**Comment by dan-lovelace** [comment]
@martingregoire - Would it be possible to see a screenshot of the DOM structure around the particular element you're trying to replace? You may omit any page contents other than the developer tab if you wish. Something like the following with the dev tools open to the elements tab and your target block of text in view (imagine I'm trying to replace the word `platform` in this case). To do this:

1. Right click the bit of text you're interested in replacing
2. Click on "Inspect"
3. The inspector should open with your target element in view
4. Grab a screenshot of the elements view - This is the section surrounded in red below

![Screenshot 2024-07-26 at 10 39 49 PM](https://github.com/user-attachments/assets/2da11251-ab6a-40d1-88d6-10ad46e50a5d)

I'm curious about surrounding elements and how they may be affecting my searching mechanisms. So far, I haven't been able to replicate your specific scenario in an isolated environment. With a screenshot of your exact layout, I'm confident a solution can be found.

**Comment by dan-lovelace** [comment]
@martingregoire - Someone reported an issue yesterday in the same ballpark as this one. It's possible this issue has been positively impacted in the new version `0.4.1`. The update is live in the stores if you'd like to upgrade and try it out.

**Comment by dan-lovelace** [comment]
Hey @martingregoire - Wanted to give an update here. No need to read everything, but there is some good info in the TLDR that might interest you.

**TLDR**: This issue should be resolved in version 0.4.2 once it hits the stores in the next day or so. I made an update to your search pattern that should work: `\b(\d+),5(?!\d)` - the notable difference being the leading word boundary. I've also written tests that replicate your environment to the best of my knowledge and they're passing with this updated pattern. Please note the build that's currently live, 0.4.1, has unfortunately broken capture groups entirely but they're fixed in 0.4.2. I'm going to close this issue for now. Don't hesitate to reach back out if you have further troubles. Thanks for bearing with me on this one.

---

## Research results

After trying to introduce `innerText` as an option for replacement methods, I've come to the conclusion it is not the right direction for this particular issue. The way text is found today is by targeting the actual [Text](https://developer.mozilla.org/en-US/docs/Web/API/Text) nodes in the DOM. These aren't visible in the inspector but imagine an element constructed like this:

```
<p>
  Lorem ipsum sit
  <a href="https://google.com">Google</a>
  Hello, world
  <a href="https://amazon.com" style="visibility: hidden;">Amazon</a>
</p>
```

Would have three main `Text` nodes considered for replacement: `Lorem ipsum sit`, `Google`, `Hello, world` and `Amazon`.

In the screenshots provided, that block of text `Hours: 7,5` exists inside a different text node than the calendar's other elements like `Monday` and `5` (day of month). This means there's no possibility of any overlap when performing the replacement.

I've written tests using as much information as I could to create a reproduction environment here: https://github.com/dan-lovelace/word-replacer-max/blob/52f9cd8d4ac32f6645a6db3f5aff5de1728a95a6/packages/shared/cypress/e2e/replace-all.cy.ts#L191-L193

It did take a little modification to the expression to get it working but `\\b(\\d+),5(?!\\d)` seems to do it. Additionally, there have been two important updates since this issue was logged:

- #10 - Major bug with missing replacements if one already occurred in the same element. This one has been live for some time and had a large impact to several areas.
- #13 - Major bug introduced to regular expression queries by the previous fix. This wouldn't have mattered at the time of logging this issue but should be kept in mind when testing version 0.4.1.
